### PR TITLE
Add support for albumArtist

### PR DESCRIPTION
**Pull request by inverse**
Resolution of #1662 

## Different artist/album artist
image

image

## Same  artist/album artist

image

image

### CODE REVIEW

File: src/core/content/connector.js [hunk 1]
```diff
@@ -163,6 +174,15 @@ function BaseConnector() {
...
 	this.getAlbum = () => $(this.albumSelector).text();
 
+	/**
+	 * Default implementation of album artist name lookup by selector.
+	 *
+	 * Override this method for more complex behaviour.
+	 *
+	 * @return {String} Song album artist
+	 */
+	this.getAlbumArtist = () => $(this.albumArtistSelector).text();
```
Comment (alexesprit) [inline src/core/content/connector.js]: I changed getters recently (108ae84ea74c821e2f70091aeec7facd4a62df42). Now they return `Util.getTextFromSelectors(selector)` instead.

Comment (inverse) [reply src/core/content/connector.js]: Rebased from master

File: src/connectors/bandcamp.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ function getArtist() {
 	if (artist === null) {
 		artist = $('.waypoint-artist-title').text().substr(3);
 	}
-	return artist;
+	return $.trim(artist);
```
Comment (alexesprit) [inline src/connectors/bandcamp.js]: Why did you add trim? It's useless because trim is done in the default metadata filter.

File: src/connectors/bandcamp.js [hunk 3]
```diff
@@ -109,4 +109,6 @@ Connector.getUniqueID = () => {
 	return null;
 };
 
+Connector.getAlbumArtist = () => $('span[itemprop="byArtist"]').text().trim() || null;
```
Comment (alexesprit) [inline src/connectors/bandcamp.js]: The same thing about trim, but also useless fallback to null.

File: src/connectors/bandcamp.js (hunk 2 above)
Comment (inverse) [reply src/connectors/bandcamp.js]: I observed different behaviour

Comment (inverse) [reply src/connectors/bandcamp.js]:
```
Web Scrobbler: {
  "track": "Proper Preaching",
  "artist": "Luca Musto",
  "album": "Proper Preaching EP",
  "albumArtist": " \n          Luca Musto\n          ",
  "uniqueID": null,
  "duration": null,
  "currentTime": null,
  "isPlaying": true,
  "trackArt": "https://f4.bcbits.com/img/a1481550943_16.jpg"
}
```

File: src/connectors/bandcamp.js (hunk 3 above)
Comment (inverse) [reply src/connectors/bandcamp.js]: Without trim I see

```
Web Scrobbler: {
  "track": "Proper Preaching",
  "artist": "Luca Musto",
  "album": "Proper Preaching EP",
  "albumArtist": " \n          Luca Musto\n          ",
  "uniqueID": null,
  "duration": null,
  "currentTime": null,
  "isPlaying": true,
  "trackArt": "https://f4.bcbits.com/img/a1481550943_16.jpg"
}
```

Comment (inverse) [reply src/connectors/bandcamp.js]: Should getTextFromSelectors do trim?

Comment (alexesprit) [reply src/connectors/bandcamp.js]: No, it should not.

albumArtist is not trimmed, because default filter is not modified to trim albumArtist.

File: src/connectors/bandcamp.js (hunk 2 above)
Comment (alexesprit) [reply src/connectors/bandcamp.js]: What do you mean? If you see **artist** field is not trimmed, it should be fixed in another place.

File: src/connectors/bandcamp.js (hunk 3 above)
Comment (inverse) [reply src/connectors/bandcamp.js]: Ahh I'll update the filter

**Comment by inverse** [comment]
Ill finish this later.

File: src/connectors/bandcamp.js [hunk 4]
```diff
@@ -1,7 +1,8 @@
 'use strict';
 
 const filter = new MetadataFilter({
-	all: MetadataFilter.removeZeroWidth
+	all: MetadataFilter.removeZeroWidth,
+	all: MetadataFilter.trim,
```
Comment (alexesprit) [inline src/connectors/bandcamp.js]: I meant the default filter defined in `filter.js`. You still have two trim calls.

**Comment by alexesprit** [comment]
I suggest to update `info.css` to disable wrapping of track field:
```css
/* Added .track class */
.artist, .album, .track {
	overflow: hidden;
	text-overflow: ellipsis;
	white-space: nowrap;
}
```

File: src/core/content/connector.js [hunk 5]
```diff
@@ -334,6 +355,10 @@ function BaseConnector() {
 			newState.track = artistTrack.track;
 		}
 
+		if (newState.artist === newState.albumArtist) {
```
Comment (inverse) [inline src/core/content/connector.js]: @alexesprit this is before the filter is applied? where would be a good place to hook up this sort of validation or do you think this is alright to push both fields when they are the same?

Comment (alexesprit) [reply src/core/content/connector.js]: Yes, in this block you compare raw data (i.e. from the page).

I think there could be cases when raw artist and raw albumArtist are different, mainly because of leading/trailing spaces. Probably, it's much better to compape right before sending the **filtered** connector state to the controller.

It could be fixed by trimming directly in the getters, but I think it's a dirty hack, and I would not recommend to go this way if there's another solution.

> or do you think this is alright to push both fields when they are the same

I don't know how Last.fm responses if we post both artist and albumArtist if they're the same. We can just test it and see how things are going. If it's okay, then this check is redundant.

Comment (inverse) [reply src/core/content/connector.js]: Dirty hacks I'm not a fan of. Added a pre-process function to act on the filtered state.

Comment (alexesprit) [reply src/core/content/connector.js]: Yes, that's better, althougth I'd move that block to `Connector.filterState` function instead of making the new one. I consider removing delete.albumArtist is a part of filtering.

File: src/popups/info.css [hunk 6]
```diff
@@ -115,3 +115,9 @@ input {
 	cursor: default !important;
 	text-decoration: none !important;
 }
+
+.artist, .album, .track {
+	overflow: hidden;
```
Comment (alexesprit) [inline src/popups/info.css]: The similar block is already defined on line 73:
```css
.artist, .album {
	overflow: hidden;
	text-overflow: ellipsis;
	white-space: nowrap;
}
```

I suggested to update that one to add `.track` class.

**Comment by alexesprit** [comment]
**Note**: This remark it applied to Bandcamp update in the PR.

As I understand #1662, the proper behavior is when:
- artist = "Various Artists";
- albumArtist = actual track artist.

But Bandcamp implementation in this PR returns "Various Artists" as an albumArtist prop, and actual track artist as an artist prop. Should it be reversed?

**Comment by inverse** [comment]
Good catch :+1:

**Comment by alexesprit** [comment]
Btw, I left some thoughts about various artists (which are related to albumArtist feature) on Bandcamp in #2006.

File: src/connectors/bandcamp.js (hunk 2 above)
Comment (alexesprit) [reply src/connectors/bandcamp.js]: I noticed the `isArtistLetious` function is broken because of raw (i.e. non-trimmed) artist passed as a function argument. If it's a reason why you added trim call, I suggest to fix `isArtistLetious` function instead.

Not sure, if we should do it in this PR, but maybe in #2006 or another.

**Comment by inverse** [comment]
Flipped the logic around to meet desired behaviour - thnx for calling that out :+1:

**Comment by inverse** [comment]
Any other feedback @alexesprit or is it good to go? I made sure that it now goes to correct URL.

**Comment by alexesprit** [comment]
@inverse I  think it's ok now, but I haven't tested this feature yet. I can do, but not today. :(

The only thing I don't know how ListenBrainz should use albumArtist info. I read ListenBrainz API docs and found no albumArtist property.

> As I understand #1662, the proper behavior is when:
> - artist = "Various Artists";
> - albumArtist = actual track artist.

Should we update ListenBrainz to use albumArtist (the actual artist of the track) as artist, if albumArtist is not empty? If no, it will scrobble "Various Artists" as an artist title for VA tracks.

**Comment by inverse** [comment]
Good point - I didn't consider such cases.  

Perhaps it should be done such that if an integration supports albumArtist than use it the way it is. Otherwise keep existing behaviour?

**Comment by alexesprit** [comment]
> Perhaps it should be done such that if an integration supports albumArtist than use it the way it is. Otherwise keep existing behaviour?

I  think we should keep the existing behavior for ListenBrainz, i.e. use albumArtist first, then artist.

**Comment by inverse** [comment]
@paramsingh - Pinging you since you originally chimed in on a PR around ListenBrainz before. 

This PR is basically adding support for albumArtist which is supported by last.fm but was wondering did Listenbrainz support something similar? 

The problem is that if we merge this - the way albumArtist works is that the artist would be in this field and in the artist field would be something like Various Artists

**Comment by alexesprit** [comment]
@inverse I asked some guy how to scrobble albumArtist, and he said Various Artists should be albumArtist. I also found some other people told the same thing.

**Comment by inverse** [comment]
@alexesprit So the flip I did was incorrect?

**Comment by alexesprit** [comment]
@inverse seems like yes. Let's wait for the comments in #1662.

**Comment by inverse** [comment]
@alexesprit it clearly is the original way I had it.

> Album Artist tags should be applied for all compilations that have different track artists. Otherwise, some music players will list each track artist and album name as a separate album. For these tracks, set the Album Artist to something like 'Various Artists' (or whatever is appropriate in your language) and apply this title consistently.

https://www.blisshq.com/music-library-management-blog/2010/10/12/how-to-use-album_artist/

**Comment by alexesprit** [comment]
@inverse now everything is fine for me. What do you think?

**Comment by inverse** [comment]
Seems good to me from testing :+1:

**Comment by alexesprit** [comment]
@inverse btw, did you find any VA compilations on other Bandcamp players, except album and song players?

**Comment by paramsingh** [comment]
Hullo, in LB, we use the `release_aritst_name` (https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/spotify_updater/spotify_read_listens.py#L99) field in `additional_info` to submit album artists to ListenBrainz. So the artist should, imo, be the actual artist of the track  for example `Kanye West, Bon Iver` while the `release_artist_name` field contains Kanye West.

**Comment by inverse** [comment]
@paramsingh thanks! I'll update the LB implementation to include that :+1:

**Comment by inverse** [comment]
@alexesprit any other feedback or are we good to go?

Refactors LB to include the `origin_url` on both nowplaying/scrobble and added the `release_artist_name`

**Comment by alexesprit** [comment]
@inverse LGTM except missing semicolons in last commit. Feel free to merge this PR if it's complete.
